src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpRenameFileToMatchTypeRefactoring.cs

```diff
@@ -0,0 +1,76 @@
+        protected override string LanguageName => LanguageNames.CSharp;
+
+        public CSharpRenameFileToMatchTypeRefactoring(VisualStudioInstanceFactory instanceFactory, ITestOutputHelper testOutputHelper)
+            : base(instanceFactory, testOutputHelper, nameof(CSharpGenerateFromUsage))
+        {
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CodeActionsMoveType)]
+        public void RenameFileToMatchType_ExistingCode()
+        {
+            var project = new ProjectUtils.Project(ProjectName);
+
+            SetUpEditor(@"class $$MismatchedClassName { }");
+
+            VisualStudio.Editor.InvokeCodeActionList();
+            VisualStudio.Editor.Verify.CodeAction("Rename file to MismatchedClassName.cs", applyFix: true);
+
+            // Ensure the file is still open in the editor, and that the file name change was made & saved
+            VisualStudio.Editor.Verify.TextContains("class MismatchedClassName { }");
+            VisualStudio.SolutionExplorer.Verify.FileContents(project, "MismatchedClassName.cs", @"class MismatchedClassName { }");
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CodeActionsMoveType)]
+        public void RenameFileToMatchType_InSubfolder()
+        {
+            var project = new ProjectUtils.Project(ProjectName);
+
+            VisualStudio.SolutionExplorer.AddFile(project, @"folder1\folder2\test.cs", open: true);
+
+            SetUpEditor(@"class $$MismatchedClassName { }");
+
+            VisualStudio.Editor.InvokeCodeActionList();
+            VisualStudio.Editor.Verify.CodeAction("Rename file to MismatchedClassName.cs", applyFix: true);
+
+            // Ensure the file is still open in the editor, and that the file name change was made & saved
+            VisualStudio.Editor.Verify.TextContains("class MismatchedClassName { }");
+            VisualStudio.SolutionExplorer.Verify.FileContents(project, @"folder1\folder2\MismatchedClassName.cs", @"class MismatchedClassName { }");
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CodeActionsMoveType)]
+        public void RenameFileToMatchType_UndoStackPreserved()
+        {
+            var project = new ProjectUtils.Project(ProjectName);
+
+            SetUpEditor(@"$$class MismatchedClassName { }");
+            VisualStudio.Editor.SendKeys("public ");
+
+            VisualStudio.Editor.InvokeCodeActionList();
+            VisualStudio.Editor.Verify.CodeAction("Rename file to MismatchedClassName.cs", applyFix: true);
+
+            // Ensure the file is still open in the editor, and that the file name change was made & saved
+            VisualStudio.Editor.Verify.CurrentLineText("public class MismatchedClassName { }");
+            VisualStudio.SolutionExplorer.Verify.FileContents(project, "MismatchedClassName.cs", @"public class MismatchedClassName { }");
+
+            // Undo just undoes text changes. The file rename is not undone, and the changes are not saved after the undo
```

Comment:
We aren't undoing the rename? That seems a bit odd. Should we be adding something to the undo stack so the rename undoes, and then you can keep undoing further after that point?

Reply:
If we do that, we should also file a bug that `right click > rename` can't be undone either.

Reply:
I've filed https://github.com/dotnet/roslyn/issues/34290 to track this, and posted my first failed attempt to https://github.com/dotnet/roslyn/pull/34291.

Reply:
https://github.com/dotnet/roslyn/pull/34291 has an explanation of why I couldn't get this working yet.

Reply:
@jasonmalinowski @jinujoseph Do you want to block this on getting Undo working?

Reply:
Isn't it a regression of a different form, because "rename file" would do an add plus a remove, which we *do* handle undo items for?

Reply:
The "RenameFileOperation" would cause the document to close (and maybe re-open), at which point I have no undo stack in VS whatsoever. Maybe something is going on that I don't see and I'm not understanding how the undo stack works when no editor is involved.

Reply:
Okay, it's supported now. :)

Reply:
Comment needs updating then?